### DISCUSSION TITLE
String Plural Forms

### DISCUSSION THREAD

**Discussion by ghost** [obsproject/obs-studio#4943] [General]
OBS does not support plural forms, meaning in some languages strings with numbers are mostly not correctly translateable.

<details><summary>More on Plural Forms</summary>

Many languages just have 2 plural forms (`one` & `many`):
one tree
many trees

E.g. Russian has more than that: `one`, `few`, `many` & `other`, all covering different ranges.

As an example, the string `Delete %1 Sources` uses the plural form `many`, and if only one source gets deleted, the string `Delete '%1'` with the plural form `one` will be used instead, excluding all languages with additional forms.

Plural Form Rules: https://unicode-org.github.io/cldr-staging/charts/latest/supplemental/language_plural_rules.html

</details>

Crowdin provides a built-in way to handle this in the editor: ![Online Editor Screenshot](https://support.crowdin.com/assets/docs_enterprise/online_editor_plural_tabs.png)
But this requires a file format supporting plural forms, like [Gettext](https://support.crowdin.com/file-formats/po/), [QT TS](https://support.crowdin.com/file-formats/ts/) or [XML](https://support.crowdin.com/file-formats/android-xml/) ➔ requires changing the way OBS parses language files.

The reason I've created this here is to discuss if this is actually something you care about and want to change and how to approach this (RFC?) before working on anything.

**Comment by dodgepong** [comment]
Honestly I never really understood why we weren't using a standardized file format to begin with. Perhaps @jp9000 can offer some more insight there.

To what extent can CrowdIn help us if we decide to convert to something more standard like Gettext? Is it possible to just take the existing translations and force it to output to a properly formatted .po file? Or does it need a .po file input first?

**Reply by ghost** [reply]
It needs a .po source file first, to which we can then add the previous translations from the .ini files.

**Reply by jp9000** [reply]
For the localization issue, I didn't know about this particular issue. As for the file format, the file format was simple enough to write without having to include complex localization libraries. A lot of GNU libraries don't tend to be fun to adopt. Setting it up seemed complicated. There's probably something out there that's better than gettext or whatever it's called. Even Qt's localization stuff was annoyingly complicated and unpleasant.

Unfortunately, changing the localization system would probably not be trivial at this point. Surely there has to be some way to word around this or perhaps exclude the fact that it's some specific amount.

**Reply by dodgepong** [reply]
I think gettext format is worth looking into. @Vainock, can you investigate what libraries we would need to use to parse and consume gettext, and what sorts of code changes we would need to make to support plural forms?

**Reply by ghost** [reply]
I don't think Gettext is the file format we should go with, let me show why:
![image](https://user-images.githubusercontent.com/39059951/124382388-7be2dd00-dcc7-11eb-8249-74125a6e254b.png)
This would be the source file.

1. At least one comment is required, otherwise the file counts as invalid.
2. An empty `msgstr` field is required for every string.

The output/translated file would look like this:
![image](https://user-images.githubusercontent.com/39059951/124382408-8ef5ad00-dcc7-11eb-9005-a7c28877f47c.png)
A lot of unnecessary information will be kept and added, including the source texts, the comments and a big header section.

---

I've found an alternative: ICU (btw, sorry for not providing this in the first message.)
ICU is a message format syntax, basically allowing the following:
```
You have deleted {count, plural,
one {# file}
other {# files}
} from your system.
```
This is also supported by Crowdin and strictly syntax-checked.
The C/C++ lib is called 'ICU4C' ('ICU 69.1'): [Documentation](https://unicode-org.github.io/icu-docs/apidoc/released/icu4c/index.html)
As I'm not really that into C++ I'm struggling to provide valuable information on how to use and implement this.

We would not have to change the file format and would only have to merge/change the strings with plural forms.

---

Some thoughts: OBS does not support multi-line .ini strings, which means the above ICU sample would need to be written like:
`You have deleted {count, plural, one {# file} other {# files}} from your system.` -> making it harder for translators to understand the syntax.

**Comment by WizardCM** [comment]
For reference, translation use the [Text Lookup Interface](https://obsproject.com/docs/reference-libobs-util-text-lookup.html) found [here](https://github.com/obsproject/obs-studio/blob/master/libobs/util/text-lookup.c).
